core/src/test/java/org/elasticsearch/search/aggregations/bucket/range/ipv4/IPv4RangeBuilderTests.java

```diff
@@ -0,0 +1,255 @@
+                assertThat(e, hasToString(containsString(expected)));
+                assertThat(e, hasToString(containsString("splitting on \"" + splitting + "\"")));
+            }
+        }
+    }
+
+    private String invalidOctetInPosition(int invalidPosition) {
+        return invalidOctetInPositions(Collections.singleton(invalidPosition));
+    }
+
+    private String invalidOctetInPosition(int invalidPosition, Supplier<String> invalidSupplier) {
+        return invalidDottedQuad(Collections.singleton(invalidPosition), invalidSupplier);
+    }
+
+    private String invalidOctetInPositions(Set<Integer> invalidPositions) {
+        return invalidDottedQuad(invalidPositions, () -> {
+            if (randomBoolean()) {
+                return Integer.toString(randomIntBetween(Integer.MIN_VALUE, -1));
+            } else {
+                return Integer.toString(randomIntBetween(256, Integer.MAX_VALUE));
+            }
+        });
+    }
+
+    private String invalidDottedQuad(Set<Integer> invalidPositions, Supplier<String> invalidSupplier) {
+        assert invalidPositions != null;
+        for (int position : invalidPositions) {
+            assert position >= 0 && position < 4;
+        }
+        String[] octets = new String[4];
+        for (int invalidPosition : invalidPositions) {
+            octets[invalidPosition] = invalidSupplier.get();
+        }
+
+        Set<Integer> validPositions = new HashSet<>(Arrays.asList(0, 1, 2, 3));
+        validPositions.removeAll(invalidPositions);
+        for (int validPosition : validPositions) {
+            octets[validPosition] = Integer.toString(randomIntBetween(0, 255));
+        }
+        return String.join(".", octets);
+    }
+
+    private int[] intToOctets(int value) {
+        int[] octets = new int[4];
+        octets[0] = (value >> 24) & 0xFF;
+        octets[1] = (value >> 16) & 0xFF;
+        octets[2] = (value >> 8) & 0xFF;
+        octets[3] = value & 0xFF;
+        return octets;
+    }
+
+    private String octetsToString(int[] octets) {
+        assert octets != null;
+        assert octets.length == 4;
+        return String.format(Locale.ROOT, "%d.%d.%d.%d", octets[0], octets[1], octets[2], octets[3]);
+    }
+
+    private String octetsToCIDR(int[] octets, int networkMask) {
+        assert octets != null;
+        assert octets.length == 4;
+        return octetsToString(octets) + "/" + networkMask;
+    }
+
+    private void assertException(String cidr, String message, String substring) {
+        assertException(cidr, message, substring, e -> true);
+    }
+
+    private void assertException(String cidr, String message, String substring, Predicate<IllegalArgumentException> assertion) {
+        try {
+            IPv4RangeBuilder.cidrMaskToMinMax(cidr);
+            fail(message);
+        } catch (IllegalArgumentException e) {
+            assertThat(cidr, e, hasToString(containsString(substring)));
+            assertTrue(assertion.test(e));
+        }
+    }
+}
```

Comment:
it shouldn't hold up the change but i find the test a bit difficult to follow.

I guess i'd recommend adding the simpler tests to the top of the file, as the randomized invalid testing is the most difficult (random test/Tuples/streams/lambas/hamcrest matchers/many helper methods). 

I'm not sure anyway that for testing invalid cases that random testing buys us so much? I would personally rather see it test like 8 or 9 explicit bad cases and assert we get the correct exception.

For valid cases on the other hand, random tests are maybe justifiable , but this can be accomplished by just wiring the first two bytes and exhaustively testing all valid masks for all addresses over that /16, which maybe makes it contained and intuitive.